pkg/loadbalancer/services.go

```diff
@@ -0,0 +1,415 @@
+
+type ServiceWriteTxn struct {
+	statedb.WriteTxn
+}
+
+var (
+	ErrServiceSourceMismatch = errors.New("service exists from different source")
+	ErrServiceConflict       = errors.New("conflict with a service with same address, but different name")
+)
+
+// RegisterHook adds a new hook to modify a service when upserted. This must only be called from
+// providers/invokes and not after starting!
+func (s *Services) RegisterHook(h ServiceHook) {
+	s.hooks = append(s.hooks, h)
+}
+
+// Services returns the service table for reading.
+// Convenience method for reducing dependencies for service writers.
+func (s *Services) Services() statedb.Table[*Service] {
+	return s.svcs
+}
+
+// Services returns the backend table for reading.
+// Convenience method for reducing dependencies for service writers.
+func (s *Services) Backends() statedb.Table[*ServiceBackend] {
+	return s.bes
+}
+
+// WriteTxn returns a write transaction against services & backends and other additional
+// tables to be used with the methods of [Services]. The returned transaction MUST be
+// Abort()'ed or Commit()'ed.
+func (s *Services) WriteTxn(extraTables ...statedb.TableMeta) ServiceWriteTxn {
+	return ServiceWriteTxn{
+		s.db.WriteTxn(s.svcs, append(extraTables, s.bes)...),
+	}
+}
+
+func (s *Services) UpsertService(txn ServiceWriteTxn, params *ServiceParams) (created bool, err error) {
+	var svc Service
+
+	existing, _, found := s.svcs.First(txn, ServiceL3n4AddrIndex.Query(params.Frontend))
+	if found {
+		// Do not allow services with same address but different names to override each other.
+		if !existing.Name.Equal(params.Name) {
+			return false, ErrServiceConflict
+		}
+
+		// Do not allow overriding a service that has been created from
+		// another source. If the service is 'Unspec'
+		if existing.Source != params.Source {
+			return false, ErrServiceSourceMismatch
+		}
+
+		// Merge the existing fields.
+		svc = *existing
+	}
+	created = !found
+
+	svc.ServiceParams = params
+	svc.Status = reconciler.StatusPending()
+
+	// Run hooks
+	for _, hook := range s.hooks {
+		hook(txn, &svc)
+	}
+
+	_, _, err = s.svcs.Insert(txn, &svc)
+	return
+}
+
+func (s *Services) DeleteService(txn ServiceWriteTxn, name ServiceName, addr L3n4Addr) (old *Service, err error) {
+	old, _, found := s.svcs.First(txn, ServiceL3n4AddrIndex.Query(addr))
+	if !found {
+		return nil, nil
+	}
+	if !old.Name.Equal(name) {
+		return nil, ErrServiceConflict
+	}
+	err = s.deleteService(txn, old)
+	return
+}
+
+func (s *Services) DeleteServicesByName(txn ServiceWriteTxn, name ServiceName, source source.Source) error {
+	svc, _, found := s.svcs.First(txn, ServiceNameIndex.Query(name))
+	if !found {
+		return statedb.ErrObjectNotFound
+	}
+	if svc.Source != source {
+		return ErrServiceSourceMismatch
+	}
+
+	return s.deleteService(txn, svc)
+}
+
+func (s *Services) deleteService(txn ServiceWriteTxn, svc *Service) error {
+	// Release references to the backends
```

Comment:
Nit: Maybe make it clear that this will also result in backend deletion
```suggestion
	// Release references to the backends and delete them if they're orphaned
```